platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheGradlePropertiesIntegrationTest.groovy

```diff
@@ -350,6 +350,118 @@ class ConfigurationCacheGradlePropertiesIntegrationTest extends AbstractConfigur
         configurationCache.assertStateLoaded()
     }
 
+    def "reuses cache when unused project property changes on disk"() {
+        buildFile """
+            tasks.register("some")
+        """
+
+        when:
+        propertiesFile.writeProperties(foo: 'one')
+        configurationCacheRun "some"
+
+        then:
+        configurationCache.assertStateStored()
+
+        when:
+        propertiesFile.writeProperties(foo: 'two')
+        configurationCacheRun "some"
+
+        then:
+        configurationCache.assertStateLoaded()
+    }
+
+    def "invalidates cache when properties file is added"() {
+        buildFile """
+            def access = project.hasProperty('foo')
+            println("Access: '\${access}'")
+            tasks.register("some")
+        """
+
+        when:
+        configurationCacheRun "some"
+
+        then:
+        configurationCache.assertStateStored()
+
+        when:
+        propertiesFile.writeProperties(foo: 'one')
+        configurationCacheRun "some"
+
+        then:
+        configurationCache.assertStateStored()
+        outputContains "configuration cache cannot be reused because Gradle property 'foo' has changed."
+    }
+
+    def "invalidates cache when builtin project property changes on disk"() {
+        buildFile """
+            tasks.register("some")
+        """
+
+        when:
+        propertiesFile.writeProperties((property): 'one')
+        configurationCacheRun "some"
+
+        then:
+        configurationCache.assertStateStored()
+
+        when:
+        propertiesFile.writeProperties((property): 'two')
+        configurationCacheRun "some"
+
+        then:
+        configurationCache.assertStateStored()
+        outputContains "configuration cache cannot be reused because Gradle property '$property' has changed."
+
+        where:
+        property << ['version', 'group', 'status', 'buildDir', 'description']
+    }
+
+    def "invalidates cache when project property used at configuration time changes on disk"() {
+        buildFile """
+            def access = project.foo
+            println("Access: '\${access}'")
+            tasks.register("some")
+        """
+
+        when:
+        propertiesFile.writeProperties(foo: 'one')
+        configurationCacheRun "some"
+
+        then:
+        configurationCache.assertStateStored()
+
+        when:
+        propertiesFile.writeProperties(foo: 'two')
+        configurationCacheRun "some"
+
+        then:
+        configurationCache.assertStateStored()
+        outputContains "configuration cache cannot be reused because Gradle property 'foo' has changed."
+    }
+
+    def "invalidates cache when system property used at configuration time changes on disk"() {
```

Comment:
💅 I would also add a test where a system property is set in the file, used only at execution time, to verify that the change to it is visible, while still getting a cache hit.